### ISSUE TITLE
animation freeze on 5.9.1

### ISSUE THREAD

**Issue by V3rg3ss3n** [sirrobzeroone/3d_armor_flyswim#17]
On minetest 5.9.1 the animation freezes and stays on the same frame. 

When I was on mintest 5.9.0 everything was all right. 

I don't know if the issue is from this side or with mintest directly.

**Comment by sotha1** [comment]
it is freezing for me too in 5.10.0

**Comment by Martructor** [comment]
Me too

**Comment by ghost** [comment]
@sirrobzeroone
In 5.11.0 still the case.

Here the changelog.
https://docs.luanti.org/about/changelog/#590--591

5.9.0 → 5.9.1
Released on 15 September 2024.
Highlights [#](https://docs.luanti.org/about/changelog/#highlights)

    Many bugfixes :)

Client / Audiovisuals [#](https://docs.luanti.org/about/changelog/#client--audiovisuals-2)

    Reverted rendering optimizations that caused anomalies with certain hardware (grorp)
    Main menu: Mapgen flags starting with “no*” are no longer shown (duplicate) (appgurueu)
    Android: Fixed input issues (grorp)
    Android: The minimap can again be proportional to the screen size (requires an up-to-date server) (grorp)
    Sneak-sumping onto a stack of 2 nodes is again possible (SmallJoker)
    Clouds no longer disappear when approaching them (sfan5)
    The CSM help dialogue now uses “.” instead of “/” (zmv7)
    Windows: disabled touchscreen auto-detection due to lack of support of the non-SDL build (default for 5.9.x) (rubenwardy)

World / Server / Environment [#](https://docs.luanti.org/about/changelog/#world--server--environment-2)

    Network improvements for slow connections (sfan5)

Script API / Modding [#](https://docs.luanti.org/about/changelog/#script-api--modding-2)

    yawsprite entities now work on OpenGL ES 2 (Android) (sfan5)
    Fixed animation not playing back on upright sprite entities on most systems (appgurueu)
    Bone overrides again return the same angles that were specified by mods to ease motion interpolation (appgurueu)
    Fixed animations not being restartable (appgurueu)
    minetest.show_formspec with an empty formspec name is no longer marked as deprecated (appgurueu)

Misc / Maintenance [#](https://docs.luanti.org/about/changelog/#misc--maintenance-2)

    macOS 12 support (sfence)
    Various fixes
    Build fixes (AMDmi3)
    Windows: removed symlink that caused issues while extracting the game archive (sfan5)

**Comment by grorp** [comment]
The regression is introduced by commit https://github.com/luanti-org/luanti/commit/dc21924f31ff52174553c4441ce8c861f842c6f9 in Luanti 5.9.1

The problem is caused by `3d_armor_flyswim` always setting the animation in [its globalstep](http://github.com/sirrobzeroone/3d_armor_flyswim/blob/f71523586d95501be826d35bea298ffda68106e2/init.lua#L109), instead of only setting it when changed, which means it's restarted every globalstep

However, the animation is not set directly, but through `player_api.set_animation`. This should avoid the issue because [that function does an early return if the new animation equals the old one](http://github.com/luanti-org/minetest_game/blob/0351c6691595c6eb88a1bea2383f6fae46a686ce/mods/player_api/api.lua#L119-L134)

It's broken anyway because [`player_api` also has a globalstep setting the animation](http://github.com/luanti-org/minetest_game/blob/0351c6691595c6eb88a1bea2383f6fae46a686ce/mods/player_api/api.lua#L192), which runs first, and the globalstep registered by `3d_armor_flyswim` only overrides the animation afterwards. This means the new animation differs from the old one, and the safeguard in `player_api.set_animation` doesn't avoid the bug.

With some logging added, this is what happens every globalstep if you walk while sneaking with `3d_armor_flyswim` enabled:

```
[globalstep]
player_api.set_animation for grrh: walk, speed 15, loop true
stack traceback:
	/mt/games/minetest_game/mods/player_api/api.lua:136: in function 'player_set_animation'
	/mt/games/minetest_game/mods/player_api/api.lua:215: in function 'globalstep'
	/mt/games/minetest_game/mods/player_api/api.lua:228: in function </mt/games/minetest_game/mods/player_api/api.lua:227>
	/mt/bin/../builtin/common/register.lua:26: in function </mt/bin/../builtin/common/register.lua:12>
player_api.set_animation for grrh: duck, speed 15, loop true
stack traceback:
	/mt/games/minetest_game/mods/player_api/api.lua:136: in function 'set_animation'
	/mt/mods/3d_armor_flyswim/init.lua:382: in function </mt/mods/3d_armor_flyswim/init.lua:109>
	/mt/bin/../builtin/common/register.lua:26: in function </mt/bin/../builtin/common/register.lua:12>
```

Setting the animation twice each globalstep (first to a different one, then resetting to the old one) was supposedly not enough to cause the bug alone because animation changes are sent delayed/batched, instead of directly when `ObjectRef:set_animation` is called

We can't just disable the globalstep in `player_api` to fix the bug, `3d_armor_flyswim` actually relies on that to reset the animation.

**Comment by wks** [comment]
Yes.  It's just like what @grorp said.  When both `player_api` and `3d_armor_flyswim` registers global step, the two globalstep routines will be executed in turn.  To the Luanti engine, it appears as if the player is frequently switching between two animations.  For example, if `player_api` thinks the player is walking, but `3d_armor_flyswim` thinks the player is swimming, the animation will oscillate between `walk` and `swim`.  Even though `swim` overrides `walk`, the engine will frequently reset the frame to the starting frame of the swimming interval, causing the player to "twitch" between the first few frames of swimming.

I managed to solve the problem in my fork.  I simply overwrote the `player_api.globalstep` field so that every time `player_api` runs the global step, it calls our function instead of its own.  I know it is monkey patching, and it may stop working if the `player_api` changed.  But it works.

```lua
player_api.globalstep = function()
	-- Do our own global step here.
end
```

See: https://github.com/wks/3d_armor_hover/commit/90ccf3e12ba3378df89b3ed79b44a7d2ab0e7f86

*Update: Oops. @grorp already made a PR using the same idea.*
